sway/input/seatop_down.c

```diff
@@ -2,21 +2,104 @@
 #include <float.h>
 #include <wlr/types/wlr_cursor.h>
 #include <wlr/types/wlr_tablet_v2.h>
+#include <wlr/types/wlr_touch.h>
 #include "sway/input/cursor.h"
 #include "sway/input/seat.h"
 #include "sway/tree/view.h"
 #include "sway/desktop/transaction.h"
 #include "log.h"
 
+struct seatop_touch_point_event {
+	double ref_lx, ref_ly;         // touch's x/y at start of op
+	double ref_con_lx, ref_con_ly; // container's x/y at start of op
+	int32_t touch_id;
+	struct wl_list link;
+};
+
 struct seatop_down_event {
 	struct sway_container *con;
 	struct sway_seat *seat;
 	struct wl_listener surface_destroy;
 	struct wlr_surface *surface;
 	double ref_lx, ref_ly;         // cursor's x/y at start of op
 	double ref_con_lx, ref_con_ly; // container's x/y at start of op
+	struct wl_list point_events;   // seatop_touch_point_event::link
 };
 
+static void handle_touch_motion(struct sway_seat *seat,
+		struct wlr_event_touch_motion *event) {
+	struct seatop_down_event *e = seat->seatop_data;
+
+	struct seatop_touch_point_event *point_event;
+	bool found = false;
+	wl_list_for_each(point_event, &e->point_events, link) {
+		if (point_event->touch_id == event->touch_id) {
+			found = true;
+			break;
+		}
+	}
+	sway_assert(found, "Unable to find matching point_event");
+
+	double moved_x = seat->touch_x - point_event->ref_lx;
+	double moved_y = seat->touch_y - point_event->ref_ly;
+	double sx = point_event->ref_con_lx + moved_x;
+	double sy = point_event->ref_con_ly + moved_y;
+
+	wlr_seat_touch_notify_motion(seat->wlr_seat, event->time_msec,
+		event->touch_id, sx, sy);
+}
+
+static void handle_touch_up(struct sway_seat *seat,
+		struct wlr_event_touch_up *event) {
+	struct seatop_down_event *e = seat->seatop_data;
+	struct seatop_touch_point_event *point_event, *tmp;
+
+	wl_list_for_each_safe(point_event, tmp, &e->point_events, link) {
+		if (point_event->touch_id == event->touch_id) {
+			wl_list_remove(&point_event->link);
+			free(point_event);
+			break;
+		}
+	}
+
+	if (wl_list_empty(&e->point_events)) {
+		seatop_begin_default(seat);
+	}
+
+	wlr_seat_touch_notify_up(seat->wlr_seat, event->time_msec, event->touch_id);
```

Comment:
Nit: should we call this before `seatop_begin_default`?

Reply:
To be honnest, I have no idea o_O